core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+import alluxio.concurrent.LockMode;
+import alluxio.resource.LockResource;
+import alluxio.resource.RefCountLockResource;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+
+/**
+ * A cache specifically designed to contain locks and will NOT evict any entries
+ * that are in use.
+ *
+ * @param <K> key to the cache
+ */
+public class LockCache<K> {
+  private static final Logger LOG = LoggerFactory.getLogger(LockCache.class);
+  private static final float DEFAULT_LOAD_FACTOR = 0.75f;
+  private static final float SOFT_LIMIT_RATIO = 0.9f;
+
+  private final Map<K, ValNode> mCache;
+  /**
+   * SoftLimit = hardLimit * softLimitRatio
+   * once the size reaches softlimit, eviction starts to happen,
+   * once the size reaches hardlimit, blocking starts to happen.
+   */
+  private final int mHardLimit;
+  private final int mSoftLimit;
+  private final Function<? super K, ? extends ReentrantReadWriteLock> mDefaultLoader;
+  private final Lock mEvictLock;
+
+  private Iterator<Map.Entry<K, ValNode>> mIterator;
+  private long mLastSizeWarningTime;
+
+  /**
+   * Constructor for a lock cache.
+   *
+   * @param defaultLoader specify a function to generate a value based on a key
+   * @param initialSize initial size of the cache
+   * @param maxSize maximum size of the cache
+   * @param concurrencyLevel concurrency level of the cache
+   */
+  public LockCache(Function<? super K, ? extends ReentrantReadWriteLock> defaultLoader,
+      int initialSize, int maxSize, int concurrencyLevel) {
+    mDefaultLoader = defaultLoader;
+    mHardLimit = maxSize;
+    mSoftLimit = (int) Math.round(SOFT_LIMIT_RATIO * maxSize);
+    mCache = new ConcurrentHashMap<>(initialSize, DEFAULT_LOAD_FACTOR, concurrencyLevel);
+    mIterator = mCache.entrySet().iterator();
+    mEvictLock = new ReentrantLock();
+    mLastSizeWarningTime = 0;
+  }
+
+  private void evictIfOverLimit() {
```

Comment:
comment on the behavior of this method. It seems like sometimes it will evict, but sometimes it won't even though eviction is necessary (another thread is doing it).

Reply:
done.